spec/classes/lib/ccreds_spec.rb

```diff
@@ -1,146 +1,198 @@
-      ) }
-    end
-    context 'when absent' do
-      let :params do
-          {
-            :ensure  => 'absent'
-          }
+
+        case v[:type]
+        when 'array'
+          kvalue = 'invalid_type'
+          errormsg = '"invalid_type" is not an array.  It looks to be a String.'
+        when 'string'
+          kvalue = ['invalid','type']
+          errormsg = '\["invalid", "type"\] is not a string.  It looks to be a Array'
+        when 'path'
+          kvalue = 'invalid/path'
+          errormsg = '"invalid/path" is not an absolute path.'
+        when 'bool'
+          kvalue = ['invalid','type']
+          errormsg = '\["invalid", "type"\] is not a boolean.  It looks to be a Array'
+        when 'hash'
+          kvalue = ['invalid','type']
+          errormsg = '\["invalid", "type"\] is not a hash.  It looks to be a Array'
+        else
+          raise "error in your spec tests - you have an unknown type value (#{v[:type]}) for parameter #{k}."
         end
-      it { should contain_package('libpam-ccreds').with(
-        'ensure'  => 'absent'
-      ) }
-    end
-    context 'when given a custom package' do
-      let :params do
-          {
-            :package  => 'libpam-magic-pkg'
-          }
+
+        let(:params) { { :"#{k}" => kvalue } }
+
+        it 'should fail' do
+          expect {
+            should contain_class('pam::lib::ccreds')
+          }.to raise_error(Puppet::Error,/#{errormsg}/)
         end
-      it { should contain_package('libpam-magic-pkg') }
+      end
     end
   end
-  context 'on a Solaris OS' do
-    let(:facts) {
-      {
-        :osfamily      => 'Solaris',
-        :kernelrelease => '5.11',
+
+  describe 'with a valid custom value specified for parameter' do
+    let :facts do
+      { :osfamily                  => 'RedHat',
+        :operatingsystemmajrelease => '7',
       }
-    }
-    context 'with no parameters' do
-      it { should raise_error(Puppet::Error, /a custom PAM ccreds module package is required for Solaris/) }
     end
-    context 'when absent' do
-      let :params do
-          {
-            :ensure  => 'absent',
-          }
-        end
-      it { should raise_error(Puppet::Error, /a custom PAM ccreds module package is required for Solaris/) }
-    end
-    context 'when absent with custom package' do
-      let :params do
-          {
-            :ensure  => 'absent',
-            :package  => 'libpam-magic-pkg',
-          }
-        end
-      it { should contain_package('libpam-magic-pkg').with(
-        'ensure'  => 'absent'
-      ) }
-    end
-    context 'when given a custom package' do
-      let :params do
-          {
-            :package  => 'libpam-magic-pkg'
-          }
-        end
-      it { should contain_package('libpam-magic-pkg') }
+
+    context 'package_ensure' do
+      let(:params) { { :package_ensure => 'absent' } }
+
+      it {
+        should contain_package('pam_lib_ccreds').with({
+          'ensure' => 'absent',
```

Comment:
i don't check for the name, because i dont actually care here. Since we are testing what happens when you set package_ensure, that's the only thing that should be in the test.